File: bindings/ruby/lib/rustpdf/document.rb
module RustPdf
  # A PDF document being authored. Call #close (or rely on GC) to free.
  class Document
    def initialize
      @ptr = Native.call("pdf_document_new")
      raise Error, "pdf_document_new returned NULL" if @ptr.null?
    end

    def close
      return if @ptr.nil? || @ptr.null?

      Native.call("pdf_document_free", @ptr)
      @ptr = nil
    end

    # ---- configuration ------------------------------------------------------

    def pdfa(level = nil)
      RustPdf.check(
        level ? Native.call("pdf_document_pdfa_level", ptr, level) : Native.call("pdf_document_pdfa", ptr)
      )
      self
    end

    def tagged
      RustPdf.check(Native.call("pdf_document_tagged", ptr))
      self
    end

    def version=(v)
      RustPdf.check(Native.call("pdf_document_set_version", ptr, v))
    end

    def default_size(width, height)
      RustPdf.check(Native.call("pdf_document_set_default_size", ptr, width, height))
      self
    end

    def info(title: nil, author: nil, subject: nil, keywords: nil, creator: nil)
      RustPdf.check(Native.call("pdf_document_set_info", ptr, title, author, subject, keywords, creator))
      self
    end

    # ---- pages + graphics ---------------------------------------------------

    def add_page(width: nil, height: nil)
      RustPdf.check(
        width && height ? Native.call("pdf_document_add_page_sized", ptr, width, height)
                        : Native.call("pdf_document_add_page", ptr)
      )
      self
    end

    def fill_rgb(r, g, b)
      RustPdf.check(Native.call("pdf_page_set_fill_rgb", ptr, r, g, b))
      self
    end

    def stroke_rgb(r, g, b)
      RustPdf.check(Native.call("pdf_page_set_stroke_rgb", ptr, r, g, b))
      self
    end

    def line_width(w)
      RustPdf.check(Native.call("pdf_page_set_line_width", ptr, w))
      self
    end

    def rect(x, y, w, h)
      RustPdf.check(Native.call("pdf_page_rect", ptr, x, y, w, h))
      self
    end

    def fill
      RustPdf.check(Native.call("pdf_page_fill", ptr))
      self
    end

    def stroke
      RustPdf.check(Native.call("pdf_page_stroke", ptr))
      self
    end

    # ---- fonts + text -------------------------------------------------------

    def add_font_file(path)
      RustPdf.out_int { |buf| Native.call("pdf_document_add_font_file", ptr, path, buf) }
    end

    def add_font(data)
      RustPdf.out_int { |buf| Native.call("pdf_document_add_font", ptr, data, data.bytesize, buf) }
    end

    def show_text(font, size, x, y, text, heading_level: 0)
      RustPdf.check(Native.call("pdf_page_show_text", ptr, font, size, x, y, text, heading_level))
      self
    end

    def paragraph(font, size, x, y, width, text, align: Align::LEFT)
      RustPdf.check(Native.call("pdf_page_paragraph", ptr, font, size, x, y, width, align, text))
      self
    end

    # ---- images -------------------------------------------------------------

    def add_image_file(path)
      RustPdf.out_int { |buf| Native.call("pdf_document_add_image_file", ptr, path, buf) }
    end

    def add_image_png(data)
      RustPdf.out_int { |buf| Native.call("pdf_document_add_image_png", ptr, data, data.bytesize, buf) }
    end

    def add_image_jpeg(data)
      RustPdf.out_int { |buf| Native.call("pdf_document_add_image_jpeg", ptr, data, data.bytesize, buf) }
    end

    def draw_image(image, x, y, w, h)
      RustPdf.check(Native.call("pdf_page_draw_image", ptr, image, x, y, w, h))
      self
    end

    def figure(image, x, y, w, h, alt)
      RustPdf.check(Native.call("pdf_page_figure", ptr, image, x, y, w, h, alt))
      self
    end

    # ---- attachments + forms ------------------------------------------------

    def attach_file(name, mime, data, relationship: Relationship::SOURCE, description: "")
      RustPdf.check(Native.call("pdf_document_attach_file", ptr, name, mime, data, data.bytesize, relationship, description))
      self
    end

    # rect = [x0, y0, x1, y1]
    def text_field(name, page, rect, value: "", size: 0.0)
      RustPdf.check(Native.call("pdf_document_text_field", ptr, name, page, rect[0], rect[1], rect[2], rect[3], value, size))
      self
    end

    def checkbox(name, page, rect, checked)
      RustPdf.check(Native.call("pdf_document_checkbox", ptr, name, page, rect[0], rect[1], rect[2], rect[3], checked ? 1 : 0))
      self
    end

    # options: Array<String>; selected: index or nil
    def dropdown(name, page, rect, options, selected: nil, size: 0.0)
      RustPdf.check(Native.call("pdf_document_dropdown", ptr, name, page, rect[0], rect[1], rect[2], rect[3],
                                options.join("\n"), selected || -1, size))
      self
    end

    # buttons: Array<[ [x0,y0,x1,y1], "export" ]>; selected: index or nil
    def radio_group(name, page, buttons, selected: nil)
      rects = buttons.flat_map { |(r, _)| r }.pack("d*")
      cstrs = buttons.map { |(_, e)| Fiddle::Pointer[e.to_s] }
      addrs = cstrs.map(&:to_i).pack("J*")
      RustPdf.check(Native.call("pdf_document_radio_group", ptr, name, page, buttons.size, rects, addrs, selected || -1))
      cstrs.clear # released after the call returned
      self
    end

    # ---- hyperlinks + bookmarks (Tier 1) ------------------------------------

    # rect = [x0, y0, x1, y1]; link to an external URI.
    def link_uri(rect, uri)
      RustPdf.check(Native.call("pdf_page_link_uri", ptr, rect[0], rect[1], rect[2], rect[3], uri))
      self
    end

    # rect = [x0, y0, x1, y1]; link to another page (optional +top+ y-offset).
    def link_to_page(rect, page_index, top: nil)
      RustPdf.check(Native.call("pdf_page_link_to_page", ptr, rect[0], rect[1], rect[2], rect[3],
                                page_index, top || 0.0, top.nil? ? 0 : 1))
      self
    end

    # Append one outline tree (a RustPdf::Bookmark). Pre-order flattened into
    # parallel arrays and emitted in a single native call.
    def add_bookmark(bookmark)
      entries = bookmark.flatten_into(0, [])
      n = entries.size
      levels = entries.map { |e| e[0] }.pack("i!*")
      pages  = entries.map { |e| e[2] }.pack("J*")
      tops   = entries.map { |e| e[3] || 0.0 }.pack("d*")
      has    = entries.map { |e| e[3].nil? ? 0 : 1 }.pack("i!*")
      cstrs  = entries.map { |e| Fiddle::Pointer[e[1].to_s] }
      titles = cstrs.map(&:to_i).pack("J*")
      RustPdf.check(Native.call("pdf_document_add_bookmarks", ptr, n, levels, titles, pages, tops, has))
      cstrs.clear # released after the call returned
      self
    end

    # ---- ZUGFeRD / Factur-X (Tier 2) ----------------------------------------

    def facturx(xml, profile: FacturxProfile::EN16931)
      RustPdf.check(Native.call("pdf_document_facturx", ptr, xml, xml.bytesize, profile))
      self
    end

    # ---- output -------------------------------------------------------------

    def page_count
      Native.call("pdf_document_page_count", ptr)
    end

    def to_bytes
      p = ptr
      RustPdf.take_bytes { |pp, pn| Native.call("pdf_document_write", p, pp, pn) }
    end

    def save(path)
      RustPdf.check(Native.call("pdf_document_save", ptr, path))
    end

    private

    def ptr
      raise Error, "operation on a closed Document" if @ptr.nil? || @ptr.null?

      @ptr
    end
  end
end

File: bindings/csharp/RustPdf/Document.cs
namespace RustPdf;

/// <summary>A PDF document being authored. Dispose to free native memory.</summary>
public sealed class Document : IDisposable
{
    private IntPtr _h;

    public Document()
    {
        Native.Init();
        _h = Native.pdf_document_new();
        if (_h == IntPtr.Zero)
            throw new PdfException(1, "pdf_document_new returned NULL");
    }

    private IntPtr H => _h != IntPtr.Zero
        ? _h
        : throw new ObjectDisposedException(nameof(Document));

    public void Dispose()
    {
        if (_h != IntPtr.Zero)
        {
            Native.pdf_document_free(_h);
            _h = IntPtr.Zero;
        }
        GC.SuppressFinalize(this);
    }

    ~Document() => Dispose();

    // ---- configuration ------------------------------------------------------

    /// <summary>Emit PDF/A-2b (or the given <paramref name="level"/>). Requires a license.</summary>
    public Document Pdfa(PdfaLevel? level = null)
    {
        Pdf.Check(level is null ? Native.pdf_document_pdfa(H) : Native.pdf_document_pdfa_level(H, (int)level));
        return this;
    }

    /// <summary>Enable the tagged/accessible structure tree. Requires a license.</summary>
    public Document Tagged()
    {
        Pdf.Check(Native.pdf_document_tagged(H));
        return this;
    }

    /// <summary>Set the PDF version (0 = 1.4, 1 = 1.5, 2 = 1.7).</summary>
    public Document SetVersion(int v)
    {
        Pdf.Check(Native.pdf_document_set_version(H, v));
        return this;
    }

    public Document SetDefaultSize(double width, double height)
    {
        Pdf.Check(Native.pdf_document_set_default_size(H, width, height));
        return this;
    }

    public Document SetInfo(string? title = null, string? author = null, string? subject = null,
        string? keywords = null, string? creator = null)
    {
        Pdf.Check(Native.pdf_document_set_info(H, title, author, subject, keywords, creator));
        return this;
    }

    // ---- pages + graphics ---------------------------------------------------

    public Document AddPage((double Width, double Height)? size = null)
    {
        Pdf.Check(size is null
            ? Native.pdf_document_add_page(H)
            : Native.pdf_document_add_page_sized(H, size.Value.Width, size.Value.Height));
        return this;
    }

    public Document SetFillRgb(double r, double g, double b)
    {
        Pdf.Check(Native.pdf_page_set_fill_rgb(H, r, g, b));
        return this;
    }

    public Document SetStrokeRgb(double r, double g, double b)
    {
        Pdf.Check(Native.pdf_page_set_stroke_rgb(H, r, g, b));
        return this;
    }

    public Document SetLineWidth(double w)
    {
        Pdf.Check(Native.pdf_page_set_line_width(H, w));
        return this;
    }

    public Document Rect(double x, double y, double w, double h)
    {
        Pdf.Check(Native.pdf_page_rect(H, x, y, w, h));
        return this;
    }

    public Document Fill()
    {
        Pdf.Check(Native.pdf_page_fill(H));
        return this;
    }

    public Document Stroke()
    {
        Pdf.Check(Native.pdf_page_stroke(H));
        return this;
    }

    // ---- fonts + text -------------------------------------------------------

    /// <summary>Register a font from a file; returns its id.</summary>
    public int AddFontFile(string path)
    {
        Pdf.Check(Native.pdf_document_add_font_file(H, path, out int id));
        return id;
    }

    /// <summary>Register a font from TrueType/OpenType bytes; returns its id.</summary>
    public int AddFont(byte[] data)
    {
        Pdf.Check(Native.pdf_document_add_font(H, data, (nuint)data.Length, out int id));
        return id;
    }

    /// <summary>Show a line of text. <paramref name="headingLevel"/> 1..6 tags it
    /// as H1..H6 (when tagged); 0 leaves it as a paragraph.</summary>
    public Document ShowText(int font, double size, double x, double y, string text, int headingLevel = 0)
    {
        Pdf.Check(Native.pdf_page_show_text(H, font, size, x, y, text, headingLevel));
        return this;
    }

    /// <summary>Lay out a wrapping paragraph in the box at <c>(x, y, width)</c>.</summary>
    public Document Paragraph(int font, double size, double x, double y, double width, string text,
        Align align = Align.Left)
    {
        Pdf.Check(Native.pdf_page_paragraph(H, font, size, x, y, width, (int)align, text));
        return this;
    }

    // ---- images -------------------------------------------------------------

    public int AddImageFile(string path)
    {
        Pdf.Check(Native.pdf_document_add_image_file(H, path, out int id));
        return id;
    }

    public int AddImagePng(byte[] data)
    {
        Pdf.Check(Native.pdf_document_add_image_png(H, data, (nuint)data.Length, out int id));
        return id;
    }

    public int AddImageJpeg(byte[] data)
    {
        Pdf.Check(Native.pdf_document_add_image_jpeg(H, data, (nuint)data.Length, out int id));
        return id;
    }

    public Document DrawImage(int image, double x, double y, double w, double h)
    {
        Pdf.Check(Native.pdf_page_draw_image(H, image, x, y, w, h));
        return this;
    }

    /// <summary>Draw a meaningful image (tagged <c>/Figure</c> with alt text).</summary>
    public Document Figure(int image, double x, double y, double w, double h, string alt)
    {
        Pdf.Check(Native.pdf_page_figure(H, image, x, y, w, h, alt));
        return this;
    }

    // ---- attachments + forms ------------------------------------------------

    public Document AttachFile(string name, string mime, byte[] data,
        AFRelationship relationship = AFRelationship.Source, string description = "")
    {
        Pdf.Check(Native.pdf_document_attach_file(
            H, name, mime, data, (nuint)data.Length, (int)relationship, description));
        return this;
    }

    public Document TextField(string name, int page, (double, double, double, double) rect,
        string value = "", double size = 0)
    {
        var (x0, y0, x1, y1) = rect;
        Pdf.Check(Native.pdf_document_text_field(H, name, (nuint)page, x0, y0, x1, y1, value, size));
        return this;
    }

    public Document Checkbox(string name, int page, (double, double, double, double) rect, bool checkedFlag)
    {
        var (x0, y0, x1, y1) = rect;
        Pdf.Check(Native.pdf_document_checkbox(H, name, (nuint)page, x0, y0, x1, y1, checkedFlag ? 1 : 0));
        return this;
    }

    public Document Dropdown(string name, int page, (double, double, double, double) rect,
        IEnumerable<string> options, int? selected = null, double size = 0)
    {
        var (x0, y0, x1, y1) = rect;
        var joined = string.Join('\n', options);
        Pdf.Check(Native.pdf_document_dropdown(
            H, name, (nuint)page, x0, y0, x1, y1, joined, selected ?? -1, size));
        return this;
    }

    public Document RadioGroup(string name, int page,
        IReadOnlyList<((double, double, double, double) Rect, string Export)> buttons, int? selected = null)
    {
        var rects = new double[buttons.Count * 4];
        var exports = new string[buttons.Count];
        for (int i = 0; i < buttons.Count; i++)
        {
            var (r, e) = buttons[i];
            rects[i * 4] = r.Item1;
            rects[i * 4 + 1] = r.Item2;
            rects[i * 4 + 2] = r.Item3;
            rects[i * 4 + 3] = r.Item4;
            exports[i] = e;
        }
        Pdf.Check(Native.pdf_document_radio_group(
            H, name, (nuint)page, (nuint)buttons.Count, rects, exports, selected ?? -1));
        return this;
    }

    // ---- hyperlinks + bookmarks (Tier 1) ------------------------------------

    /// <summary>Add a clickable link rectangle that opens an external URI.</summary>
    public Document LinkUri((double, double, double, double) rect, string uri)
    {
        var (x0, y0, x1, y1) = rect;
        Pdf.Check(Native.pdf_page_link_uri(H, x0, y0, x1, y1, uri));
        return this;
    }

    /// <summary>Add a clickable link rectangle that jumps to another page
    /// (optionally scrolled to <paramref name="top"/>).</summary>
    public Document LinkToPage((double, double, double, double) rect, int pageIndex, double? top = null)
    {
        var (x0, y0, x1, y1) = rect;
        Pdf.Check(Native.pdf_page_link_to_page(
            H, x0, y0, x1, y1, (nuint)pageIndex, top ?? 0.0, top is null ? 0 : 1));
        return this;
    }

    /// <summary>Append one outline (bookmark) tree to the document. Children are
    /// flattened pre-order and submitted in a single native call.</summary>
    public Document AddBookmark(Bookmark bookmark)
    {
        var entries = new List<(int Level, string Title, int Page, double? Top)>();
        bookmark.Flatten(0, entries);
        int n = entries.Count;
        var levels = new int[n];
        var titles = new string[n];
        var pages = new nuint[n];
        var tops = new double[n];
        var hasTops = new int[n];
        for (int i = 0; i < n; i++)
        {
            var e = entries[i];
            levels[i] = e.Level;
            titles[i] = e.Title;
            pages[i] = (nuint)e.Page;
            hasTops[i] = e.Top is null ? 0 : 1;
            tops[i] = e.Top ?? 0.0;
        }
        Pdf.Check(Native.pdf_document_add_bookmarks(H, (nuint)n, levels, titles, pages, tops, hasTops));
        return this;
    }

    // ---- ZUGFeRD / Factur-X (Tier 2) ----------------------------------------

    /// <summary>Embed a Factur-X/ZUGFeRD invoice XML (PDF/A-3). Requires a license.</summary>
    public Document Facturx(byte[] xml, FacturxProfile profile = FacturxProfile.En16931)
    {
        Pdf.Check(Native.pdf_document_facturx(H, xml, (nuint)xml.Length, (int)profile));
        return this;
    }

    // ---- output -------------------------------------------------------------

    public int PageCount => Native.pdf_document_page_count(H);

    public byte[] ToBytes() => Pdf.TakeBuffer((out IntPtr p, out nuint n) => Native.pdf_document_write(H, out p, out n));

    public void Save(string path) => Pdf.Check(Native.pdf_document_save(H, path));
}
